File: equadiff.f90
!Ce code Fortran modelise le mouvement d'une planete autour du soleil. GM=1, M>>m

!Execution:
!gfortran equadiff.f90 -o equadiff.x

!Gnuplot:
!plot 'rk4.out' using 2:3
!set xlabel 'x'
!set ylabel 'y'
!replot

!_________________________

!X(1)=x(t)
!X(2)=y(t)
!X(3)=vx(t)
!X(4)=vy(t)

!Xderiv(1)=vx(t)
!Xderiv(2)=vy(t)
!Xderiv(3)=-x/(r^3)
!Xderiv(4)=-y/(r^3)

!conditions initiales annoncees dans l'enonce // 
!x0=0 // x0=1
!y0=1 // y0=0
!vx0=-0.5 // vx0=0
!vy0=0 // vy0=-0.5

program  meca_planet
implicit none
real (8) :: dt ,t ,ener
integer  :: i
real(8),  dimension (4) :: X, Xderiv
external  :: euler, deriv_planet, rk4

t=0.
dt=0.01

!Initialization
X=(/1.,0.,0.,-0.5/)

open(11,file='rk4.out')

do i=1,100000!tmax=0.01*100000=1000
	t=t+dt
	call  rk4(t,X,dt,4,deriv_planet)
	ener =0.5*(X(3)**2+X(4)**2)-1/sqrt(X(1)**2+X(2)**2)
	if (mod(nint(t/dt),10).eq.0) then
		write(11,*) t, X!, ener
	endif
enddo
close (11)
end program meca_planet

subroutine deriv_planet(t,X,Xderiv,n)
implicit none
integer, intent(in) :: n
real(8), intent (in) :: t
real(8) :: xnew, ynew, radius
real(8), dimension(n), intent(in) :: X
real(8), dimension(n), intent(out) :: Xderiv
if (n.ne.4) write (*,*) 'WARNING: dimension de n incorrecte, devrait etre 4'
radius=sqrt(X(1)**2+X(2)**2)
Xderiv(1)=X(3)
Xderiv(2)=X(4)
Xderiv(3)=-X(1)/radius**3
Xderiv(4)=-X(2)/radius**3
end subroutine deriv_planet

subroutine rk4(t,X,dt,n,deriv)
!Runge-Kutta du 4eme ordre
implicit none
integer, intent(in) :: n
real(8), intent(in) :: t, dt
real(8), dimension(n), intent (inout) :: X
real(8) :: ddt
real(8), dimension(n) :: Xp, k1, k2, k3, k4
ddt = 0.5*dt
call deriv(t,X,k1,n); Xp=X+ddt*k1
call deriv(t+ddt,Xp,k2,n); Xp=X+ddt*k2
call deriv(t+ddt,Xp,k3,n); Xp=X+dt*k3
call deriv(t+dt,Xp,k4,n); X = X + dt*(k1+2.0*k2 + 2.0*k3+k4)/6.0
end subroutine rk4

File: equadiff.cpp
/*
!Ce code C++ modelise le mouvement d'une planete autour du soleil. GM=1, M>>m

Compilation:
c++ -Wall -Wextra equadiff.cpp -o equadiffcpp.x

Temps pris par la fonction principale: 297211 microsecondes, soit 0.297211 seconde

Gnuplot:
plot 'equadiffrk4cpp.res' using 2:3
set xlabel 'x'
set ylabel 'y'
replot

_________________________

X(1)=x(t)
X(2)=y(t)
X(3)=vx(t)
X(4)=vy(t)

Xderiv(1)=vx(t)
Xderiv(2)=vy(t)
Xderiv(3)=-x/(r^3)
Xderiv(4)=-y/(r^3)

conditions initiales annoncees dans l'enonce // utilisees ici
x0=0 // x0=1
y0=1 // y0=0
vx0=-0.5 // vx0=0
vy0=0 // vy0=-0.5
*/

#include<iostream>
#include<fstream>
#include<cmath>
#include <chrono> //pour mesurer le temps d'execution

using namespace std;
using namespace std::chrono;

void deriv(double t, double X[], double Xderiv[], int n){
double radius;
radius=sqrt(pow(X[0],2)+pow(X[1],2));
Xderiv[0]=X[2];
Xderiv[1]=X[3];
Xderiv[2]=-X[0]/pow(radius,3);
Xderiv[3]=-X[1]/pow(radius,3);
}



void rk4(double t, double X[], double dt, int n, void deriv(double, double[], double[], int)){

int i;
double ddt;
double Xp[n], k1[n], k2[n], k3[n], k4[n];
ddt=0.5*dt;

deriv(t,X,k1,n);

for(i=0;i<n;i++){
Xp[i]=X[i]+ddt*k1[i];
deriv(t+ddt,Xp,k2,n);
}

for(i=0;i<n;i++){
Xp[i]=X[i]+ddt*k2[i];
deriv(t+ddt,Xp,k3,n);
}

for(i=0;i<n;i++){
Xp[i]=X[i]+dt*k3[i];
deriv(t+dt,Xp,k4,n);
}

for(i=0;i<n;i++){
X[i] = X[i] + dt*(k1[i]+2*k2[i]+2*k3[i]+k4[i])/6;
}
}


int main(void){
double dt, t, tmax;
double X[4];
double Xderiv[4];

dt=0.01;
tmax=1000.0;

X[0]=1.0;
X[1]=0.0;
X[2]=0.0;
X[3]=-0.5;

ofstream fichier ("equadiffrk4cpp.out");

auto start = high_resolution_clock::now();//on commence a compter le temps de mesure

for(t=0.0;t<=tmax;t+=dt){
	rk4(t,X,dt,4,deriv);
	if((int)(round(t/dt))%10==0){//on n'ecrit qu'une valeur sur 10
	fichier <<t<<" "<<X[0]<<" "<<X[1]<<endl;
	}
}

auto stop = high_resolution_clock::now();//on arrete de compter le temps d'execution

fichier.close();

auto duration = duration_cast<microseconds>(stop - start); 
  
cout << "Time taken by function: "
         << duration.count() << " microseconds" << endl; 
  
return 0; 

}

